server/sonar-server/src/test/java/org/sonar/server/tester/UserSessionRule.java

```diff
@@ -0,0 +1,318 @@
+import org.sonar.server.user.UserSession;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * {@code UserSessionRule} is intended to be used as a {@link org.junit.Rule} to easily manage {@link UserSession} in
+ * unit tests.
+ * <p>
+ * It can be used as a {@link org.junit.ClassRule} but be careful not to modify its states from inside tests methods
+ * unless you purposily want to have side effects between each tests.
+ * </p>
+ * <p>
+ * {@code UserSessionRule} is intended to be used either standalone (in which case use the static factory method
+ * {@link #standalone()} or with {@link ServerTester} (in which case use static factory method
+ * {@link #forServerTester(ServerTester)}).
+ * </p>
+ * <p>
+ * In both cases, one can define user session behavior which should apply on all tests directly on the property, eg.:
+ * <pre>
+ * {@literal @}Rule
+ * public UserSessionRule userSessionRule = UserSessionRule.standalone().logon("admin").setGlobalPermissions(GlobalPermissions.SYSTEM_ADMIN);
+ * </pre>
+ * </p>
+ * <p>
+ * Behavior defined at property-level can obviously be override at test method level. For example, one could define
+ * all methods to use an authenticated session such as presented above but can easily overwrite that behavior in a
+ * specific test method as follow:
+ * <pre>
+ * {@literal @}Test
+ * public void test_method() {
+ *   userSessionRule.standalone();
+ *   {@literal [...]}
+ * }
+ * </pre>
+ * </p>
+ * <p>
+ * {@code UserSessionRule}, being standalone or associated to a {@link ServerTester}, emulates by default an anonymous
+ * session. Therefore, call {@code UserSessionRule.standalone()} is equivalent to calling
+ * {@code UserSessionRule.standalone().anonymous()}.
+ * </p>
+ * <p>
+ * To emulate an identified user, either use method {@link #logon(String)} if you want to specify the user's login, or
+ * method {@link #logon()} which will do the same but using the value of {@link #DEFAULT_LOGIN} as the user's login.
+ * </p>
+ */
+public class UserSessionRule implements TestRule, UserSession {
+  public static final String DEFAULT_LOGIN = "default_login";
+
+  @CheckForNull
+  private final ServerTester serverTester;
+  private UserSession currentUserSession;
+
+  private UserSessionRule(@Nullable ServerTester serverTester) {
+    this.serverTester = serverTester;
+    anonymous();
+  }
+
+  public static UserSessionRule standalone() {
+    return new UserSessionRule(null);
+  }
+
+  public static UserSessionRule forServerTester(ServerTester serverTester) {
+    return new UserSessionRule(Preconditions.checkNotNull(serverTester));
+  }
+
+  /**
+   * Log on with the default login {@link #DEFAULT_LOGIN}
+   */
+  public UserSessionRule logon() {
+    return logon(DEFAULT_LOGIN);
+  }
+
+  /**
+   * Log on with the specified login
+   */
+  public UserSessionRule logon(String login) {
```

Comment:
I would prefer that too, but someone decided that getter convention wasn't right and decided to use `login` instead of `getLogin` as the name of the method to get the login of the current `UserSession`. I used `logon` to avoid unsolvable conflict and avoid confusion.

btw: isn't there a rule about method named exactly the same as a property? or is that only in findbug and/or disabled on SQ?